x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/stats/StatsDirectoryWrapper.java

```diff
@@ -0,0 +1,260 @@
+            final long filePointer = getFilePointer();
+
+            final byte readByte = input.readByte();
+            if (readByte != -1) {
+                stats.incrementBytesRead(lastReadPosition.getAndUpdate(l -> getFilePointer()) == filePointer);
+            }
+            return readByte;
+        }
+
+        @Override
+        public void readBytes(byte[] b, int offset, int len) throws IOException {
+            final long filePointer = getFilePointer();
+            input.readBytes(b, offset, len);
+            stats.incrementBytesRead(lastReadPosition.getAndUpdate(l -> getFilePointer()) == filePointer, len);
+        }
+
+        @Override
+        public void seek(final long pos) throws IOException {
+            final long filePointer = getFilePointer();
+            input.seek(pos);
+            stats.incrementSeekCount(pos - filePointer);
+        }
+
+        @Override
+        public IndexInput clone() {
+            final StatsIndexInputWrapper clone = new StatsIndexInputWrapper(input.clone(), stats.incrementCloneCount());
+            assert this.stats == clone.stats : "same live stats object should be propagated to clones";
+            clone.isClone = true;
+            return clone;
+        }
+
+        @Override
+        public IndexInput slice(String sliceDesc, long off, long len) throws IOException {
+            final StatsIndexInputWrapper slice = new StatsIndexInputWrapper(input.slice(sliceDesc, off, len), stats.incrementSliceCount());
+            assert this.stats == slice.stats : "same live stats object should be propagated to slices";
+            slice.isClone = true;
+            return slice;
+        }
+    }
+
+    /**
+     * {@link LiveStats} records stats for a given {@link IndexInput} and all its clones and slices.
+     */
+    private static class LiveStats {
+
+        final LongAdder opened = new LongAdder();
+        final LongAdder closed = new LongAdder();
+        final LongAdder sliced = new LongAdder();
+        final LongAdder cloned = new LongAdder();
+
+        final LiveCounter forwardSeeks = new LiveCounter();
+        final LiveCounter backwardSeeks = new LiveCounter();
+
+        final LiveCounter contiguousReads = new LiveCounter();
+        final LiveCounter nonContiguousReads = new LiveCounter();
+
+        LiveStats incrementOpenCount() {
+            opened.increment();
+            return this;
+        }
+
+        LiveStats incrementCloseCount() {
+            closed.increment();
+            return this;
+        }
+
+        LiveStats incrementSliceCount() {
+            sliced.increment();
+            return this;
+        }
+
+        LiveStats incrementCloneCount() {
+            cloned.increment();
+            return this;
+        }
+
+        void incrementSeekCount(final long n) {
+            if (n >= 0) {
+                forwardSeeks.add(n);
+            } else {
+                backwardSeeks.add(n != Long.MIN_VALUE ? -n : 0L);
```

Comment:
> I wonder if we should just `backwardSeeks.add(n);` for simplicity, and deal with the negative total elsewhere.

Agreed.

> I also think it might be useful to distinguish forward seeks by size. One possibility would be to set a threshold to distinguish large and small seeks.

Yes. I added the length of the index input to the stats (it is useful to interpret the stats) and I made the distinction between small and large seeks (I picked < 25% of length as a threshold)